subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,540 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_img_grp
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+#include <pm_config.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h>
+
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#include <mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h>
+
+/* Mimimal number of ZCBOR encoder states to provide full encoder functionality plus one for backup
+ */
+#define CBOR_ENCODER_STATE_NUM 3
+
+#define MCUMGR_IMAGE_LIST_SIZE 2
+
+struct upload_progress {
+	char sha256[IMG_MGMT_HASH_LEN];
+	bool hash_initialized;
+	size_t image_size;
+	size_t offset;
+	size_t request_length;
+	size_t writed_length;
+	int image_num;
+};
+
+static struct upload_progress upload_state;
+static struct mcumgr_image_state *image_info;
+static struct mcumgr_image_upload *image_upload_buf;
+static int image_erase_status;
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static const char smp_images_str[] = "images";
+#define IMAGES_STR_LEN (sizeof(smp_images_str) - 1)
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size)
+{
+	for (int i = 0; i < map_size; i++) {
+		/* Clear found status */
+		map[i].found = false;
+	}
+}
+
+static int image_state_res_fn(struct net_buf *nb, void *user_data)
+{
+	zcbor_state_t zsd[CBOR_ENCODER_STATE_NUM + 2];
+	struct zcbor_string value = {0};
+	int buf_len, slot_num, img_num, rc;
+	struct zcbor_string hash, version;
+	bool bootable, pending, confirmed, active, permanent, ok;
+	size_t decoded;
+	struct zcbor_map_decode_key_val list_res_decode[] = {
+		/** Mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("version", zcbor_tstr_decode, &version),
+		ZCBOR_MAP_DECODE_KEY_DECODER("hash", zcbor_bstr_decode, &hash),
+		ZCBOR_MAP_DECODE_KEY_DECODER("slot", zcbor_int32_decode, &slot_num),
+		/** End of mandatory components */
```

Comment:
Don't need /** for comments here because there files are not documented, replace the "end of" with "optional", these aren't components, they're fields